src/cpp/session/modules/SessionRVersions.cpp

```diff
@@ -0,0 +1,123 @@
+/*
+ * SessionRVersions.cpp
+ *
+ * Copyright (C) 2009-19 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include "SessionRVersions.hpp"
+
+#include <core/Exec.hpp>
+#include <session/prefs/UserPrefs.hpp>
+#include <session/SessionModuleContext.hpp>
+#include <session/RVersionSettings.hpp>
+
+#include <r/RExec.hpp>
+#include <r/ROptions.hpp>
+#include <r/RRoutines.hpp>
+#include <r/RJson.hpp>
+
+using namespace rstudio::core;
+using namespace rstudio::session::prefs;
+
+namespace rstudio {
+namespace session {
+namespace modules {
+namespace r_versions {
+namespace {
+
+void syncRVersionPref()
+{
+   RVersionSettings versionSettings(module_context::userScratchPath(),
+                                 FilePath(options().getOverlayOption(
+                                             kSessionSharedStoragePath)));
+
+   auto value = prefs::userPrefs().readValue(kDefaultRVersion);
+   if (!value || value->type() != json::ObjectType)
```

Comment:
Should we warn if the value exists but is not an object type? Or will that happen for us automagically with schema validation?

Reply:
It happens with schema validation; the warning will be emitted then -- so no need to double warn here. However `get_obj` throws for non-object types so this is an extra safeguard.